README.md
# bebop_vservo
Visual servo controller for Parrot Bebop Drone (ROS package)


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package bebop_vservo
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.1.0 (2016-02-12)
------------------
* Major refactoring and changes
  - First robust version
  - Yaw is directly being controlled via image error (requires vel_ctrl
  for yaw)
  - Subscribe to bebop_yaw (not used for now)
  - Simpler logic
* Contributors: Mani Monajjemi

1.0.0 (2016-02-11)
------------------
* Initial stable version
* Contributors: Mani Monajjemi


src/bebop_vservo_nodelet.cpp
// c++
#include <cmath>

// boost
#include <boost/thread/lock_guard.hpp>

// ros
#include <pluginlib/class_list_macros.h>
#include <angles/angles.h>
#include <visp_bridge/camera.h>

// visp
#include <visp/vpFeatureBuilder.h>

#include "bebop_vservo/bebop_vservo_nodelet.h"

namespace bebop_vservo
{

BebopVServoNodelet::BebopVServoNodelet()
  : enabled_(false),
    servo_inited(false),
    roi_recv_time(ros::Time(0)),
    fov_x(0.0),
    fov_y(0.0),
    cam_tilt_rad(0.0),
    t_start_loop(0)

{
}

BebopVServoNodelet::~BebopVServoNodelet()
{
  // nodelet log macros do not work here!

  if (ctrlloop_thread_ptr_)
  {
    ctrlloop_thread_ptr_->interrupt();
    ctrlloop_thread_ptr_->join();
  }
  std::cout << "[CVER] Graceful shutdown" << std::endl;
}

void BebopVServoNodelet::UpdateParams()
{
  util::GetParam(private_nh_, "desired_depth", depth, 2.5);
  util::GetParam(private_nh_, "start_enabled", enabled_, false);
  util::GetParam(private_nh_, "servo_gain", lambda, 0.4);
  util::GetParam(private_nh_, "update_freq", update_freq, 30.0);
  util::GetParam(private_nh_, "height_target", height_target_m, 0.5);
  util::GetParam(private_nh_, "distground_target", distground_target_m, 0.75);
  util::GetParam(private_nh_, "cam_tilt", cam_tilt_rad, 0.0);
  cam_tilt_rad = angles::from_degrees(cam_tilt_rad);
}

void BebopVServoNodelet::EnableCallback(const std_msgs::BoolConstPtr &enable_msg_ptr)
{
  boost::lock_guard<boost::mutex> lock(mutex_enable_);
  enabled_ptr_ = enable_msg_ptr;
  ROS_INFO_STREAM("[VSER] Enable request: " << enable_msg_ptr->data);
}

void BebopVServoNodelet::RoiCallback(const sensor_msgs::RegionOfInterestConstPtr& roi_msg_ptr)
{
  boost::lock_guard<boost::mutex> lock(mutex_roi_);
  roi_cptr_ = roi_msg_ptr;
  roi_recv_time = ros::Time::now();
}

void BebopVServoNodelet::CameraOrientationCallback(const bebop_msgs::Ardrone3CameraStateOrientationConstPtr& cam_ori_ptr)
{
  boost::lock_guard<boost::mutex> lock(mutex_servo);
  cam_tilt_rad = -angles::from_degrees(cam_ori_ptr->tilt);
  NODELET_INFO_STREAM("[VSER] Bebop camera's new tilt: " << -cam_ori_ptr->tilt);
}

void BebopVServoNodelet::CameraCallback(const sensor_msgs::ImageConstPtr& img_ptr,
                                        const sensor_msgs::CameraInfoConstPtr& cinfo_msg_ptr)
{
  {
    boost::lock_guard<boost::mutex> lock(mutex_servo);
    cam_model.fromCameraInfo(cinfo_msg_ptr);
    fov_x = 2.0 * atan2(cam_model.fullResolution().width / 2.0, cam_model.fx());
    fov_y = 2.0 * atan2(cam_model.fullResolution().height/ 2.0, cam_model.fy());
    NODELET_INFO_STREAM_ONCE("[VSER] FOV " << angles::to_degrees(fov_x) << " " << angles::to_degrees(fov_y));
  }

  bool is_valid_roi = false;
  sensor_msgs::RegionOfInterest roi_cached;
  {
    boost::lock_guard<boost::mutex> lock(mutex_roi_);
    is_valid_roi = (ros::Time::now() - roi_recv_time).toSec() < (5.0 / update_freq);
    if (is_valid_roi && roi_cptr_) roi_cached = *roi_cptr_;
  }

  {
    boost::lock_guard<boost::mutex> lock(mutex_servo);
    if (!servo_inited && is_valid_roi)
    {
      cam = visp_bridge::toVispCameraParameters(*cinfo_msg_ptr);
      cam.printParameters();
      //lambda_adapt.initStandard(4.0, 0.4, 40.0);

      vpFeatureBuilder::create(p[0], cam, vpImagePoint(roi_cached.y_offset, roi_cached.x_offset));
      vpFeatureBuilder::create(p[1], cam, vpImagePoint(roi_cached.y_offset, roi_cached.x_offset + roi_cached.width));
      vpFeatureBuilder::create(p[2], cam, vpImagePoint(roi_cached.y_offset + roi_cached.height, roi_cached.x_offset + roi_cached.width));
      vpFeatureBuilder::create(p[3], cam, vpImagePoint(roi_cached.y_offset + roi_cached.height, roi_cached.x_offset));

      vpPoint point[4];
      const double tw = height_target_m / 2.0;
      point[0].setWorldCoordinates(-tw, -tw, 0);
      point[1].setWorldCoordinates( tw, -tw, 0);
      point[2].setWorldCoordinates( tw,  tw, 0);
      point[3].setWorldCoordinates(-tw,  tw, 0);

      vpHomogeneousMatrix cMo;
      vpTranslationVector cto(0, 0, depth);
      vpRxyzVector cro(vpMath::rad(0.0), vpMath::rad(0.0), vpMath::rad(0.0));
      vpRotationMatrix cRo(cro);
      cMo.buildFrom(cto, cRo);

      for (uint32_t i = 0; i < 4; i++)
      {
        vpColVector cP, p_img;
        point[i].changeFrame(cMo, cP);
        point[i].projection(cP, p_img);

        pd[i].set_xyZ(p_img[0], p_img[1], cP[2]);
      }

      for (uint32_t i = 0; i < 4; i++)
      {
        task.addFeature(p[i], pd[i]);
      }

      task.setServo(vpServo::EYEINHAND_CAMERA);
      task.setInteractionMatrixType(vpServo::MEAN, vpServo::PSEUDO_INVERSE);
      task.setLambda(lambda);
      task.print();
      servo_inited = true;
    }

    if (!servo_inited)
    {
      NODELET_WARN_THROTTLE(1, "[VSER] Wating for the first roi ...");
    }
  }
}
void BebopVServoNodelet::onInit()
{
  nh_ = getNodeHandle();
  private_nh_ = getPrivateNodeHandle();
  UpdateParams();
  util::ResetCmdVel(cmd_vel);

  it_ptr_ = boost::make_shared<image_transport::ImageTransport>(nh_);

  // In Nodelet version, the default transport hint does not use Nodelet's local and
  // private Nodehandles
  image_transport::TransportHints it_th("raw", ros::TransportHints(), private_nh_);

  camera_sub = it_ptr_->subscribeCamera("bebop/image_raw", 1, &BebopVServoNodelet::CameraCallback, this);

  NODELET_INFO_STREAM("[VSER] Image transport in use: " << camera_sub.getTransport());

  cam_orientation_sub = nh_.subscribe("bebop/states/ARDrone3/CameraState/Orientation", 10
                                       , &BebopVServoNodelet::CameraOrientationCallback, this);

  roi_sub = nh_.subscribe("track_roi", 1, &BebopVServoNodelet::RoiCallback, this);
  enable_sub = nh_.subscribe("enable", 1, &BebopVServoNodelet::EnableCallback, this);

  cmd_vel_pub = nh_.advertise<geometry_msgs::Twist>("cmd_vel", 10);

  ctrlloop_thread_ptr_ = boost::make_shared<boost::thread>(
        boost::bind(&BebopVServoNodelet::ControlLoopThread, this));
}

void BebopVServoNodelet::Reset()
{
  util::ResetCmdVel(cmd_vel);
  cmd_vel_pub.publish(cmd_vel);
}

void BebopVServoNodelet::ControlLoopThread()
{
  NODELET_INFO("[VSER] Control loop thread started ...");
  ros::Rate loop_rate(update_freq);
  while (!boost::this_thread::interruption_requested())
  {
    try
    {
      if (!loop_rate.sleep())
      {
        throw std::runtime_error("[VSER] Missed target frequency");
      }

      {
        boost::lock_guard<boost::mutex> lock(mutex_enable_);
        if (enabled_ptr_)
        {
          enabled_ = enabled_ptr_->data;
        }
      }

      if (!enabled_)
      {
        ROS_WARN_THROTTLE(1, "[VSER] Not enabled ...");
        continue;
      }

      bool is_valid_roi = false;
      sensor_msgs::RegionOfInterest roi;
      {
        boost::lock_guard<boost::mutex> lock(mutex_roi_);
        is_valid_roi = (ros::Time::now() - roi_recv_time).toSec() < (5.0 / update_freq);
        if (is_valid_roi && roi_cptr_) roi = *roi_cptr_;
        if (roi.width < 5 || roi.height < 5 ||
            (roi.x_offset + roi.width >= cam_model.fullResolution().width) ||
            (roi.y_offset + roi.width >= cam_model.fullResolution().height))
        {
          is_valid_roi = false;
        }
      }
      if (!is_valid_roi)
      {
        ROS_WARN_THROTTLE(1, "[VSER] ROI is too old or not valid");
        Reset();
        continue;
      }

      {
        boost::lock_guard<boost::mutex> lock(mutex_servo);
        const double im_width_px = static_cast<double>(cam_model.fullResolution().width);
        const double im_height_px = static_cast<double>(cam_model.fullResolution().height);
        double sigma_1 = static_cast<double>(roi.y_offset) / im_height_px * fov_y;
        double sigma_2 = (im_height_px - static_cast<double>(roi.y_offset + roi.height)) / im_height_px * fov_y;
        double roi_height_px = roi.height;
        double beta_rad = (roi_height_px / im_height_px) * fov_y;
        double d1_m = height_target_m * sin(M_PI_2 - fov_y/2.0 + sigma_2 - cam_tilt_rad) / sin(beta_rad);
        double z1_m = d1_m * sin(M_PI_2 + fov_y/2.0  - cam_tilt_rad - sigma_1);
        vpFeatureBuilder::create(p[0], cam, vpImagePoint(roi.y_offset, roi.x_offset));
        vpFeatureBuilder::create(p[1], cam, vpImagePoint(roi.y_offset, roi.x_offset + roi.width));
        vpFeatureBuilder::create(p[2], cam, vpImagePoint(roi.y_offset + roi.height, roi.x_offset + roi.width));
        vpFeatureBuilder::create(p[3], cam, vpImagePoint(roi.y_offset + roi.height, roi.x_offset));
        for (uint32_t i = 0; i < 4; i++)
        {
           p[i].set_Z(z1_m);
        }
        v = task.computeControlLaw();
        task.print();
        ROS_WARN_STREAM("V_SERVO\n" << v);
      }
    }
    catch (const std::runtime_error& e)
    {
      ROS_ERROR_STREAM("[VSER] Runtime error " << e.what());
      Reset();
    }
  }

  NODELET_INFO("[VSER] Control loop thread stopped ...");
}
}  // namespace bebop_servo

PLUGINLIB_DECLARE_CLASS(bebop_vservo, BebopVServoNodelet, bebop_vservo::BebopVServoNodelet, nodelet::Nodelet);


include/bebop_vservo/bebop_vservo_ctrl.h
#ifndef BEBOP_VSERVO_CTRL_H
#define BEBOP_VSERVO_CTRL_H

// boost
#include <boost/shared_ptr.hpp>

// ros
#include <ros/node_handle.h>
#include <sensor_msgs/CameraInfo.h>
#include <sensor_msgs/RegionOfInterest.h>
#include <geometry_msgs/Twist.h>
#include <image_geometry/pinhole_camera_model.h>
#include <std_msgs/Bool.h>

#include "bebop_msgs/Ardrone3PilotingStateAttitudeChanged.h"
#include "bebop_msgs/Ardrone3CameraStateOrientation.h"

// visp && visp_bridge
#include <visp/vpServo.h>
#include <visp/vpFeaturePoint.h>
#include <visp/vpCameraParameters.h>


// bebop_vservo
#include "bebop_vservo/Debug.h"
#include "bebop_vservo/Target.h"

#ifndef CLAMP
#define CLAMP(x, l, h) (((x) > (h)) ? (h) : (((x) < (l)) ? (l) : (x)))
#endif

namespace bebop_vservo
{
namespace util {

inline void ResetCmdVel(geometry_msgs::Twist& v)
{
  v.linear.x = 0.0;
  v.linear.y = 0.0;
  v.linear.z = 0.0;
  v.angular.x = 0.0;
  v.angular.y = 0.0;
  v.angular.z = 0.0;
}

template<typename T>
bool GetParam(const ros::NodeHandle& nh, const::std::string& key, T& val)
{
  if (nh.getParam(key, val))
  {
    ROS_INFO_STREAM("[VSER] Param " << key << " : " << val);
    return true;
  }
  ROS_WARN_STREAM("[VSER] Param " << key << " not found/set.");
  return false;
}

template<typename T>
bool GetParam(const ros::NodeHandle& nh, const::std::string& key, T& val, const T& default_val)
{
  nh.param(key, val, default_val);
  ROS_INFO_STREAM("[VSER] Param " << key << " : " << val);
}

}  // namespace util

class BebopVServoCtrl
{
public:
  explicit BebopVServoCtrl(ros::NodeHandle &nh);
  virtual void Spin();

private:
  bool enabled_;
  bool caminfo_recv_;
  bool servo_inited_;
  bool force_reinit_;

  ros::NodeHandle nh_;
  ros::NodeHandle nh_priv_;

  ros::Subscriber sub_caminfo_;
  ros::Subscriber sub_enable_;
  ros::Subscriber sub_roi_;
  ros::Subscriber sub_cam_orientation_;
  ros::Subscriber sub_bebop_att_;

  ros::Publisher pub_debug_;
  ros::Publisher pub_cmd_vel_;

  geometry_msgs::Twist msg_cmd_vel_;
  bebop_vservo::Debug msg_debug_;

  bebop_vservo::TargetConstPtr target_cptr_;

  // Internal
  ros::Time target_recv_time_;
  ros::Time bebop_recv_time_;

  double fov_x_;
  double fov_y_;
  image_geometry::PinholeCameraModel cam_model_;
  double cam_tilt_rad_;
  double bebop_yaw_rad_;

  // Visp
  boost::shared_ptr<vpServo> vp_task_ptr_;

  //vpFeaturePoint fpd_[4];
  //vpFeaturePoint fp_[4];

  std::vector<vpFeaturePoint> fpd_;
  std::vector<vpFeaturePoint> fp_;

  vpCameraParameters vp_cam_;
  double vp_gain_;
  vpColVector vp_v_;
  vpAdaptiveGain vp_gain_adapt_;

  // These are set by the user when target is first received or when reinit=true
  double servo_desired_depth_;
  double servo_target_height_;
  double servo_target_width_;
  double servo_target_dist_ground_;
  double servo_desired_yaw_rad_;

  // params
  double param_update_freq_;

  void Reset();
  bool Update();

  void BebopAttitudeCallback(const bebop_msgs::Ardrone3PilotingStateAttitudeChangedConstPtr& att_ptr);
  void CameraOrientationCallback(const bebop_msgs::Ardrone3CameraStateOrientationConstPtr& cam_ori_ptr);
  void CameraCallback(const sensor_msgs::CameraInfoConstPtr& msg_ptr);
  void TargetCallback(const bebop_vservo::TargetConstPtr target_msg_ptr);
  void EnableCallback(const std_msgs::BoolConstPtr& enable_msg_ptr);

  void UpdateParams();
};

}  // namespace bebop_vservo
#endif


include/bebop_vservo/bebop_vservo_nodelet.h
#ifndef BEBOP_VSERVO_NODELET_H
#define BEBOP_VSERVO_NODELET_H

// boost
#include <boost/thread.hpp>
#include <boost/shared_ptr.hpp>

// ros
#include <ros/node_handle.h>
#include <nodelet/nodelet.h>
#include <image_transport/image_transport.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/CameraInfo.h>
#include <sensor_msgs/RegionOfInterest.h>
#include <geometry_msgs/Twist.h>
#include <image_geometry/pinhole_camera_model.h>
#include <std_msgs/Bool.h>

#include "bebop_msgs/Ardrone3CameraStateOrientation.h"

// visp && visp_bridge
#include <visp/vpServo.h>
#include <visp/vpFeaturePoint.h>
#include <visp/vpCameraParameters.h>


// bebop_vservo
#include "bebop_vservo/debug.h"

#ifndef CLAMP
#define CLAMP(x, l, h) (((x) > (h)) ? (h) : (((x) < (l)) ? (l) : (x)))
#endif

namespace bebop_vservo
{
namespace util {

inline void ResetCmdVel(geometry_msgs::Twist& v)
{
  v.linear.x = 0.0;
  v.linear.y = 0.0;
  v.linear.z = 0.0;
  v.angular.x = 0.0;
  v.angular.y = 0.0;
  v.angular.z = 0.0;
}

template<typename T>
bool GetParam(const ros::NodeHandle& nh, const::std::string& key, T& val)
{
  if (nh.getParam(key, val))
  {
    ROS_INFO_STREAM("[VSER] Param " << key << " : " << val);
    return true;
  }
  ROS_WARN_STREAM("[VSER] Param " << key << " not found/set.");
  return false;
}

template<typename T>
bool GetParam(const ros::NodeHandle& nh, const::std::string& key, T& val, const T& default_val)
{
  nh.param(key, val, default_val);
  ROS_INFO_STREAM("[VSER] Param " << key << " : " << val);
}

}  // namespace util

class BebopVServoNodelet: public nodelet::Nodelet
{
public:
  BebopVServoNodelet();
  virtual ~BebopVServoNodelet();

private:
  std_msgs::BoolConstPtr enabled_ptr_;
  bool enabled_;
  bool servo_inited;

  boost::shared_ptr<boost::thread> ctrlloop_thread_ptr_;

  boost::mutex mutex_enable_;
  boost::mutex mutex_roi_;
  boost::mutex mutex_servo;

  ros::NodeHandle nh_;
  ros::NodeHandle private_nh_;
  boost::shared_ptr<image_transport::ImageTransport> it_ptr_;
  ros::Subscriber enable_sub;
  ros::Subscriber roi_sub;
  ros::Subscriber cam_orientation_sub;
  image_transport::CameraSubscriber camera_sub;

  ros::Publisher debug_pub;
  ros::Publisher cmd_vel_pub;

  geometry_msgs::Twist cmd_vel;
  bebop_vservo::debug debug_msg;
  sensor_msgs::RegionOfInterestConstPtr roi_cptr_;

  // Internal
  ros::Time roi_recv_time;
  double fov_x;
  double fov_y;
  image_geometry::PinholeCameraModel cam_model;
  double cam_tilt_rad;

  // Visp
  vpServo task;
  vpFeaturePoint pd[4];
  vpFeaturePoint p[4];
  vpCameraParameters cam;
  double Z, Zd;
  double lambda;
  double t_start_loop;
  double tinit;
  vpColVector v;
  vpColVector v_beb;
  vpColVector vi;
  double mu;
  vpAdaptiveGain lambda_adapt;

  // params
  double depth;
  double height_target_m;
  double distground_target_m;
  double update_freq;

  void Reset();
  void ControlLoopThread();
  void CameraOrientationCallback(const bebop_msgs::Ardrone3CameraStateOrientationConstPtr& cam_ori_ptr);
  void CameraCallback(const sensor_msgs::ImageConstPtr& img_ptr, const sensor_msgs::CameraInfoConstPtr& cinfo_msg_ptr);
  void RoiCallback(const sensor_msgs::RegionOfInterestConstPtr& roi_msg_ptr);
  void EnableCallback(const std_msgs::BoolConstPtr& enable_msg_ptr);

  void UpdateParams();

protected:
  virtual void onInit();

};

}  // namespace bebop_vservo
#endif


src/bebop_vservo_ctrl.cpp
// c++
#include <cmath>

// boost
#include <boost/make_shared.hpp>

// ros
#include <angles/angles.h>
#include <visp_bridge/camera.h>

// visp
#include <visp/vpFeatureBuilder.h>

#include "bebop_vservo/bebop_vservo_ctrl.h"

namespace bebop_vservo
{

BebopVServoCtrl::BebopVServoCtrl(ros::NodeHandle &nh)
  : enabled_(false),
    servo_inited_(false),
    caminfo_recv_(false),
    force_reinit_(false),
    nh_(nh),
    nh_priv_("~"),
    sub_caminfo_(nh.subscribe("bebop/camera_info", 1, &BebopVServoCtrl::CameraCallback, this)),
    sub_cam_orientation_(nh_.subscribe("bebop/states/ARDrone3/CameraState/Orientation", 10
                                      , &BebopVServoCtrl::CameraOrientationCallback, this)),
    sub_bebop_att_(nh_.subscribe("bebop/states/ARDrone3/PilotingState/AttitudeChanged", 10,
                                 &BebopVServoCtrl::BebopAttitudeCallback, this)),
    sub_roi_(nh_.subscribe("target", 1, &BebopVServoCtrl::TargetCallback, this)),
    sub_enable_(nh_.subscribe("enable", 1, &BebopVServoCtrl::EnableCallback, this)),
    pub_cmd_vel_(nh_.advertise<geometry_msgs::Twist>("cmd_vel", 10)),
    pub_debug_(nh_.advertise<bebop_vservo::Debug>("debug", 30)),
    target_recv_time_(ros::Time(0)),
    bebop_recv_time_(ros::Time(0)),
    fov_x_(0.0),
    fov_y_(0.0),
    cam_tilt_rad_(0.0),
    bebop_yaw_rad_(0.0),
    vp_task_ptr_()  // set to null
{
  UpdateParams();
  util::ResetCmdVel(msg_cmd_vel_);
}

void BebopVServoCtrl::UpdateParams()
{
  util::GetParam(nh_priv_, "start_enabled", enabled_, false);
  util::GetParam(nh_priv_, "servo_gain", vp_gain_, 0.4);
  util::GetParam(nh_priv_, "update_freq", param_update_freq_, 30.0);
  util::GetParam(nh_priv_, "cam_tilt", cam_tilt_rad_, 0.0);
  cam_tilt_rad_ = angles::from_degrees(cam_tilt_rad_);
}

void BebopVServoCtrl::EnableCallback(const std_msgs::BoolConstPtr &enable_msg_ptr)
{
  enabled_ = enable_msg_ptr->data;
  ROS_INFO_STREAM("[VSER] request: " << (enabled_ ? "enable" : "disable"));
}

void BebopVServoCtrl::TargetCallback(const TargetConstPtr target_msg_ptr)
{
  target_cptr_ = target_msg_ptr;
  force_reinit_ = static_cast<bool>(target_msg_ptr->reinit);
  target_recv_time_ = ros::Time::now();
}

void BebopVServoCtrl::CameraOrientationCallback(const bebop_msgs::Ardrone3CameraStateOrientationConstPtr& cam_ori_ptr)
{
  const double& new_tilt_angle_rad = -angles::from_degrees(cam_ori_ptr->tilt);
  if (fabs(cam_tilt_rad_ - new_tilt_angle_rad) > 0.01)
  {
    // TEMP
    ROS_DEBUG_STREAM("[VSER] Bebop camera's new tilt: " << -cam_ori_ptr->tilt);
  }
  cam_tilt_rad_ = new_tilt_angle_rad;
}

void BebopVServoCtrl::BebopAttitudeCallback(const bebop_msgs::Ardrone3PilotingStateAttitudeChangedConstPtr &att_ptr)
{
  bebop_yaw_rad_ = -att_ptr->yaw;
  bebop_recv_time_ = ros::Time::now();
}

void BebopVServoCtrl::CameraCallback(const sensor_msgs::CameraInfoConstPtr& cinfo_msg_ptr)
{
  if (!caminfo_recv_)
  {
    cam_model_.fromCameraInfo(cinfo_msg_ptr);
    fov_x_ = 2.0 * atan2(cam_model_.fullResolution().width / 2.0, cam_model_.fx());
    fov_y_ = 2.0 * atan2(cam_model_.fullResolution().height/ 2.0, cam_model_.fy());
    ROS_INFO_STREAM_ONCE("[VSER] FOV " << angles::to_degrees(fov_x_) << " " << angles::to_degrees(fov_y_));

    vp_cam_ = visp_bridge::toVispCameraParameters(*cinfo_msg_ptr);
    vp_cam_.printParameters();

    caminfo_recv_ = true;
    // Might be an overkill
    ROS_WARN("[VSER] Camera info received, shutting down the subscriber.");
    sub_caminfo_.shutdown();
  }

}

void BebopVServoCtrl::Reset()
{
  util::ResetCmdVel(msg_cmd_vel_);
  pub_cmd_vel_.publish(msg_cmd_vel_);
}

void BebopVServoCtrl::Spin()
{
  ROS_INFO("[VSER] Control loop started ...");

  ros::Rate loop_rate(param_update_freq_);
  while (ros::ok())
  {
    try
    {
      ros::spinOnce();
      if (!loop_rate.sleep())
      {
        throw std::runtime_error("[VSER] Missed target frequency");
      }

      if (!enabled_)
      {
        ROS_WARN_THROTTLE(10, "[VSER] Not enabled ...");
        continue;
      }
      if (!Update()) Reset();
    }
    catch (const vpException& e)
    {
      ROS_ERROR_STREAM("[VSER] Visp error " << e.what());
      Reset();
    }
    catch (const ros::Exception& e)
    {
      ROS_ERROR_STREAM("[VSER] ROS error " << e.what());
      Reset();
    }
    catch (const std::runtime_error& e)
    {
      ROS_ERROR_STREAM("[VSER] Runtime error " << e.what());
      Reset();
    }
  }
}

bool BebopVServoCtrl::Update()
{
  if (!caminfo_recv_)
  {
    ROS_WARN("[VSER] The camera info has not been received yet!");
    return false;
  }

  //bool is_valid_roi = (ros::Time::now() - target_recv_time_).toSec() < (5.0 / param_update_freq_);
  bool is_valid_roi = (ros::Time::now() - target_recv_time_).toSec() < 2.0;

  sensor_msgs::RegionOfInterest roi;
  if (is_valid_roi && target_cptr_) roi = target_cptr_->roi;
  if (roi.width < 5 || roi.height < 5 ||
      (roi.x_offset + roi.width >= cam_model_.fullResolution().width) ||
      (roi.y_offset + roi.height >= cam_model_.fullResolution().height))
  {
    is_valid_roi = false;
  }

  if (!is_valid_roi)
  {
    ROS_WARN_THROTTLE(1, "[VSER] ROI is too old or not valid. Disabling visual servo!");
    // Experimental
    enabled_ = false;
    servo_inited_ = false;
    return false;
  }

  if (is_valid_roi && (!servo_inited_ || force_reinit_))
  {
    // This will re-set all internal params of the task
    vp_task_ptr_ = boost::make_shared<vpServo>();
    fp_.clear();
    fpd_.clear();
    fp_.resize(4);
    fpd_.resize(4);


    servo_inited_ = false;

    servo_desired_depth_ = target_cptr_->desired_depth;
    servo_desired_yaw_rad_ = target_cptr_->desired_yaw_rad;
    servo_target_height_ = target_cptr_->target_height_m;
    servo_target_width_ = target_cptr_->target_width_m;
    servo_target_dist_ground_ = target_cptr_->target_distance_ground;

    ROS_WARN_STREAM("[VSER] (re-)init " <<
                    " is_valid_roi: " << (is_valid_roi ? "yes" : "no") <<
                    " re-init request: " << (force_reinit_ ? "yes" : "no") <<
                    " Desired depth: " << servo_desired_depth_ <<
                    " Desired yaw: " << angles::to_degrees(servo_desired_yaw_rad_) <<
                    " Target height: " << servo_target_height_ <<
                    " Target width: " << servo_target_width_ <<
                    " Target d2 ground: " << servo_target_dist_ground_
                    );

    ROS_WARN_STREAM("[VSER] Target ROI [x,y,w,h]: " << roi.x_offset << " " << roi.y_offset << " " << roi.width << " " << roi.height);

    //lambda_adapt.initStandard(4.0, 0.4, 40.0);

    ROS_ASSERT(fp_.size() == 4);
    vpFeatureBuilder::create(fp_[0], vp_cam_, vpImagePoint(roi.y_offset, roi.x_offset));
    vpFeatureBuilder::create(fp_[1], vp_cam_, vpImagePoint(roi.y_offset, roi.x_offset + roi.width));
    vpFeatureBuilder::create(fp_[2], vp_cam_, vpImagePoint(roi.y_offset + roi.height, roi.x_offset + roi.width));
    vpFeatureBuilder::create(fp_[3], vp_cam_, vpImagePoint(roi.y_offset + roi.height, roi.x_offset));

    vpPoint point[4];
    const double h2 = servo_target_height_ / 2.0;
    const double w2 = servo_target_width_ / 2.0;
    point[0].setWorldCoordinates(-w2, -h2, 0);
    point[1].setWorldCoordinates( w2, -h2, 0);
    point[2].setWorldCoordinates( w2,  h2, 0);
    point[3].setWorldCoordinates(-w2,  h2, 0);

//    point[0].setWorldCoordinates(-h2, -w2, 0);
//    point[1].setWorldCoordinates( h2, -w2, 0);
//    point[2].setWorldCoordinates( h2,  w2, 0);
//    point[3].setWorldCoordinates(-h2,  w2, 0);

    vpHomogeneousMatrix cMo;
    vpTranslationVector cto(0, 0, servo_desired_depth_);
    vpRxyzVector cro(vpMath::rad(0.0), vpMath::rad(0.0), vpMath::rad(0.0));
    vpRotationMatrix cRo(cro);
    cMo.buildFrom(cto, cRo);

    for (uint32_t i = 0; i < 4; i++)
    {
      vpColVector cP, p_img;
      point[i].changeFrame(cMo, cP);
      point[i].projection(cP, p_img);

      fpd_[i].set_xyZ(p_img[0], p_img[1], cP[2]);
    }

    ROS_ASSERT(vp_task_ptr_);
    ROS_ASSERT((fp_.size() == 4) && (fpd_.size() == 4));

    for (uint32_t i = 0; i < 4; i++)
    {
      vp_task_ptr_->addFeature(fp_[i], fpd_[i]);
    }

    vp_task_ptr_->setServo(vpServo::EYEINHAND_CAMERA);
    vp_task_ptr_->setInteractionMatrixType(vpServo::MEAN, vpServo::PSEUDO_INVERSE);
    vp_task_ptr_->setLambda(vp_gain_);
    vp_task_ptr_->print();
    servo_inited_ = true;
    force_reinit_ = false;
  }

  if (!servo_inited_ || !vp_task_ptr_)
  {
    ROS_WARN_THROTTLE(1, "[VSER] Servo task has not been initialized yet.");
    return false;
  }

  const double im_width_px = static_cast<double>(cam_model_.fullResolution().width);
  const double im_height_px = static_cast<double>(cam_model_.fullResolution().height);
  double sigma_1 = static_cast<double>(roi.y_offset) / im_height_px * fov_y_;
  double sigma_2 = (im_height_px - static_cast<double>(roi.y_offset + roi.height)) / im_height_px * fov_y_;
  double roi_height_px = roi.height;
  double beta_rad = (roi_height_px / im_height_px) * fov_y_;
  double d1_m = servo_target_height_ * sin(M_PI_2 - fov_y_/2.0 + sigma_2 - cam_tilt_rad_) / sin(beta_rad);
  double z1_m = d1_m * sin(M_PI_2 + fov_y_/2.0  - cam_tilt_rad_ - sigma_1);

  ROS_ASSERT(fp_.size() == 4);

  vpFeatureBuilder::create(fp_[0], vp_cam_, vpImagePoint(roi.y_offset, roi.x_offset));
  vpFeatureBuilder::create(fp_[1], vp_cam_, vpImagePoint(roi.y_offset, roi.x_offset + roi.width));
  vpFeatureBuilder::create(fp_[2], vp_cam_, vpImagePoint(roi.y_offset + roi.height, roi.x_offset + roi.width));
  vpFeatureBuilder::create(fp_[3], vp_cam_, vpImagePoint(roi.y_offset + roi.height, roi.x_offset));

  for (uint32_t i = 0; i < 4; i++)
  {
    fp_[i].set_Z(z1_m);
  }
  vp_v_ = vp_task_ptr_->computeControlLaw();
  //vp_task_.print();
  ROS_DEBUG_STREAM("V_SERVO: " << vp_v_.transpose());

  /* EXPERIMENTAL */
//  ROS_ERROR_STREAM("[VSER] delay: " << (ros::Time::now() - bebop_recv_time_).toSec());
//  ROS_ERROR_STREAM("[VSER] current yaw err (deg): " << angles::to_degrees(angles::normalize_angle(servo_desired_yaw_rad_ - bebop_yaw_rad_)));

  const double yaw_image_err_raw =
      ((cam_model_.fullResolution().width/2.0 - (roi.x_offset + roi.width / 2.0)) / (double) cam_model_.fullResolution().width) * fov_x_;

  vp_v_[0] = 0.0;
  servo_desired_yaw_rad_ = vp_gain_ * yaw_image_err_raw;
  /* END */


  // Here we should convert from Camera (Visp) coordinates to Bebop Coordinates
  // TODO: Consider camera tilt
  // Servo:           z: forward, x: right, y: down
  // Maps to bebop:   z: x_b    , x: -y_b , y: -z_b

  util::ResetCmdVel(msg_cmd_vel_);

  // No clamp/filter here
  // Map to Bebop's frame from camera's frame

  const double& vx_cam = vp_v_[2];
  const double& vz_cam = -vp_v_[1];

  // TODO: Check the freshness of tilt!
  msg_cmd_vel_.linear.x =  vx_cam * cos(cam_tilt_rad_) + vz_cam * sin(cam_tilt_rad_);
  msg_cmd_vel_.linear.z = -vx_cam * sin(cam_tilt_rad_) + vz_cam * cos(cam_tilt_rad_);
  msg_cmd_vel_.linear.y = -vp_v_[0];

//  msg_cmd_vel_.linear.x = vp_v_[2];
//  msg_cmd_vel_.linear.y = -vp_v_[0];
//  msg_cmd_vel_.linear.z = -vp_v_[1];

  // We can't control all four DOF of the vehicle
  msg_cmd_vel_.angular.z = servo_desired_yaw_rad_;

  pub_cmd_vel_.publish(msg_cmd_vel_);

  msg_debug_.header.stamp = ros::Time::now();
  msg_debug_.header.frame_id = "";
  msg_debug_.bb_height = roi_height_px;
  //msg_debug_.d_groundtruth = d_truth;
  msg_debug_.cam_tilt = angles::to_degrees(cam_tilt_rad_);
  msg_debug_.beta = beta_rad;
  msg_debug_.sigma_1 = sigma_1;
  msg_debug_.sigma_2 = sigma_2;
  msg_debug_.d_raw_bb = z1_m;
  msg_debug_.target_height = servo_target_height_;
  msg_debug_.target_grounddist = servo_target_dist_ground_;
  for (uint32_t i = 0; i < 6; ++i)
    msg_debug_.v_img[i] = vp_v_.data[i];

  msg_debug_.desired_depth = servo_desired_depth_;
  msg_debug_.desired_yaw_rad = servo_desired_yaw_rad_;
  msg_debug_.target_roi = roi;

  pub_debug_.publish(msg_debug_);
  return true;
}

}  // namespace bebop_servo
